internal/overlord/logstate/syslog.go

```diff
@@ -0,0 +1,279 @@
+
+	canonicalPrivEnterpriseNum = 28978
+)
+
+type syslogForwarder struct {
+	serviceName string
+	targetName  string
+	target      plan.LogTarget
+
+	iterator servicelog.Iterator
+	parser   *servicelog.Parser
+	stopCh   chan struct{}
+
+	buffer *syslogBuffer
+	client *syslogClient
+}
+
+func newSyslogForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*syslogForwarder, error) {
+	client, err := newSyslogClient(target.Location)
+	if err != nil {
+		return nil, err
+	}
+
+	buffer := newSyslogBuffer()
+	return &syslogForwarder{
+		serviceName: serviceName,
+		targetName:  targetName,
+		target:      target,
+		iterator:    iterator,
+		parser:      servicelog.NewParser(iterator, lokiParserSize),
+		stopCh:      make(chan struct{}, 1),
+		buffer:      buffer,
+		client:      client,
+	}, nil
+}
+
+func (f *syslogForwarder) Forward() {
+	for {
+		// Check if Stop has been called
+		select {
+		case <-f.stopCh:
+			f.flushBuffer()
+			return
+		default:
+			// continue
+		}
+
+		if !f.buffer.IsEmpty() {
+			f.flushBuffer()
+		}
+
+		if f.parser.Next() {
+			f.buffer.WriteEntry(f.parser.Entry())
+		} else {
+			if err := f.parser.Err(); err != nil {
+				logger.Noticef(
+					"Forwarding logs to target %q: cannot parse log for service %q: %v",
+					f.targetName, f.serviceName, err)
+			}
+		}
+	}
+}
+
+func (f *syslogForwarder) Stop() {
+	f.stopCh <- struct{}{}
+}
+
+func (f *syslogForwarder) flushBuffer() {
+	err := f.client.Send(&f.buffer.data)
+	if err != nil {
+		logger.Noticef(
+			"Forwarding logs to target %q: cannot send logs for service %q: %v",
+			f.targetName, f.serviceName, err)
+		return
+	}
+}
+
+type syslogBuffer struct {
+	version        int
+	host           string
+	msgid          string
+	priority       int
+	pid            string
+	structuredData string // TODO: use this for future log labels
+
+	data bytes.Buffer
+}
+
+func newSyslogBuffer() *syslogBuffer {
+	return &syslogBuffer{
+		version:        1,
+		pid:            "-",
+		host:           "-",
+		msgid:          "-",
+		priority:       priorityVal(FacilityUserLevelMessage, SeverityInformational),
+		structuredData: syslogStructuredData("pebble", canonicalPrivEnterpriseNum, nil),
+	}
+}
+
+// Syslog Priority values - see RFC 5424 6.2.1
+const (
+	// Facility values
+	FacilityUserLevelMessage = 1
+
+	// Severity values
+	SeverityInformational = 6
+)
+
+// priorityVal calculates the syslog Priority value (PRIVAL) from the given
+// Facility and Severity values. See RFC 5424, sec 6.2.1 for details.
+func priorityVal(facility, severity int) int {
+	return facility*8 + severity
+}
+
+// syslogStructuredData formats the given labels into a structured data section for a syslog message
+// according to RFC5424 section 6.
+func syslogStructuredData(name string, enterpriseNum int, labels map[string]string) string {
+	if len(labels) == 0 {
+		return "-"
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", name, enterpriseNum)
```

Comment:
I was going to comment on efficiency, but given this is only done once, this is fine -- thanks!